includes/data-port/class-sensei-data-port-job.php

```diff
@@ -0,0 +1,332 @@
+	 *     }
+	 * }
+	 *
+	 * @var array
+	 */
+	private $logs;
+
+	/**
+	 * True if the job is completed.
+	 *
+	 * @var bool
+	 */
+	private $is_completed;
+
+	/**
+	 * True if the internal state has changed and needs to be stored.
+	 *
+	 * @var bool
+	 */
+	private $has_changed;
+
+	/**
+	 * Estimate of completion percentage.
+	 *
+	 * @var float
+	 */
+	private $percentage;
+
+	/**
+	 * The tasks that consist this data port job.
+	 *
+	 * @var Sensei_Data_Port_Task_Interface[]
+	 */
+	private $tasks;
+
+	/**
+	 * Sensei_Data_Port_Job constructor.
+	 *
+	 * @param string $job_id   Unique job id.
+	 * @param string $json     A json string to restore internal state from.
+	 */
+	protected function __construct( $job_id, $json = '' ) {
+		$this->job_id      = $job_id;
+		$this->has_changed = false;
+
+		if ( '' !== $json ) {
+			$this->restore_from_json( $json );
+		} else {
+			$this->logs         = [];
+			$this->results      = [];
+			$this->is_completed = false;
+			$this->has_changed  = true;
+			$this->task_state   = [];
+			$this->percentage   = 0;
+		}
+
+		$this->tasks = $this->get_tasks();
+
+		add_action( 'shutdown', [ $this, 'persist' ] );
+	}
+
+	/**
+	 * Restore a stored job. Returns null if the job does not exist.
+	 *
+	 * @param string $job_id  The job id.
+	 *
+	 * @return Sensei_Data_Port_Job|null instance.
+	 */
+	public static function get( $job_id ) {
+		$json = get_option( self::get_option_name(), '' );
+
+		if ( empty( $json ) ) {
+			return null;
+		}
+
+		return new static( $job_id, $json );
+	}
+
+	/**
+	 * Get the results of the job.
+	 *
+	 * @return array The results.
+	 */
+	public function get_results() {
+		return $this->results;
+	}
+
+	/**
+	 * Get the logs of the job. The logs are grouped by type. The pagination works on the total number of logs which
+	 * means that depending on the arguments, multiple type of logs can be returned.
+	 *
+	 * @param int $offset  Offset for pagination.
+	 * @param int $limit   Limit for pagination.
+	 *
+	 * @return array The logs.
+	 */
+	public function get_logs( $offset = 0, $limit = 20 ) {
+
+		$result_logs = [];
+
+		foreach ( $this->logs as $type => $messages ) {
+
+			// Check if the logs should start from this group.
+			if ( $offset < count( $messages ) ) {
+				$added_logs           = array_slice( $messages, $offset, $limit );
+				$offset               = 0;
+				$result_logs[ $type ] = $added_logs;
+
+				if ( count( $added_logs ) >= $limit ) {
+					return $result_logs;
+				} else {
+					// Adjust the limit to take into account added logs.
+					$limit -= count( $added_logs );
+				}
+			} else {
+				// Adjust the offset for the skipped log entries.
+				$offset -= count( $messages );
+			}
+		}
+
+		return $result_logs;
+	}
+
+	/**
+	 * Delete any stored state for this job.
+	 */
+	public function clean_up() {
+		foreach ( $this->tasks as $task ) {
+			$task->cleanup();
+		}
+
+		delete_option( self::get_option_name( $this->job_id ) );
```

Comment:
I wonder if we need to do anything to prevent this from accidentally persisting again on shutdown. Maybe just setting `has_changed` to `false` is enough.

Reply:
Good catch! 14e581f3403efc82f2dbafde49889cf563692024

Reply:
I ended up adding a new variable since using `had_changed` woudln't work in all cases. b8b4ca4c2a692aca9cf4fe86275b2eb1bf61f8cf